pkg/template/event_test.go

```diff
@@ -0,0 +1,309 @@
+		Name:  "paramNoEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar"},
+	}
+	wantParamNoEventPathVar = pipelinev1.Param{
+		Name:  "paramNoEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar"},
+	}
+	paramOneEventPathVar = pipelinev1.Param{
+		Name:  "paramOneEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event.one)-bar"},
+	}
+	wantParamOneEventPathVar = pipelinev1.Param{
+		Name:  "paramOneEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-onevalue-bar"},
+	}
+	paramMultipleIdenticalEventPathVars = pipelinev1.Param{
+		Name:  "paramMultipleIdenticalEventPathVars",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event.one)-$(event.one)-$(event.one)-bar"},
+	}
+	wantParamMultipleIdenticalEventPathVars = pipelinev1.Param{
+		Name:  "paramMultipleIdenticalEventPathVars",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-onevalue-onevalue-onevalue-bar"},
+	}
+	paramMultipleUniqueEventPathVars = pipelinev1.Param{
+		Name:  "paramMultipleUniqueEventPathVars",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event.one)-$(event.two.two)-$(event.three.three.three.three.three)-bar"},
+	}
+	wantParamMultipleUniqueEventPathVars = pipelinev1.Param{
+		Name:  "paramMultipleUniqueEventPathVars",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-onevalue-twovalue-threevalue-bar"},
+	}
+	paramSubobjectEventPathVar = pipelinev1.Param{
+		Name:  "paramSubobjectEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event.three)-bar"},
+	}
+	wantParamSubobjectEventPathVar = pipelinev1.Param{
+		Name:  "paramSubobjectEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: `bar-{"three": {"three": {"three": {"three": "threevalue"}}}}-bar`},
+	}
+	paramEntireEventEventPathVar = pipelinev1.Param{
+		Name:  "paramEntireEventEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event)-bar"},
+	}
+	wantParamEntireEventEventPathVar = pipelinev1.Param{
+		Name:  "paramEntireEventEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: `bar-{"one": "onevalue", "two": {"two": "twovalue"}, "three": {"three": {"three": {"three": {"three": "threevalue"}}}}}-bar`},
+	}
+	paramOneBogusEventPathVar = pipelinev1.Param{
+		Name:  "paramOneBogusEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event.bogus.path)-bar"},
+	}
+	wantParamOneBogusEventPathVar = pipelinev1.Param{
+		Name:  "paramOneBogusEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event.bogus.path)-bar"},
+	}
+	paramMultipleBogusEventPathVars = pipelinev1.Param{
+		Name:  "paramMultipleBogusEventPathVars",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event.bogus.path)-$(event.two.bogus)-$(event.three.bogus)-bar"},
+	}
+	wantParamMultipleBogusEventPathVars = pipelinev1.Param{
+		Name:  "paramMultipleBogusEventPathVars",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event.bogus.path)-$(event.two.bogus)-$(event.three.bogus)-bar"},
```

Comment:
So we will always opt to use String from the pipelinev1.ArrayOrString, but people could potentially still inject whatever the `$(event)` binding resolves to, which could be an array?

Reply:
Yep, that's my intention. I'm thinking whatever is injected, we will handle as a string (at least for now). This simplifies things as we get the code working for our first iteration. Also, JSON objects support [these types](https://www.w3schools.com/js/js_json_datatypes.asp), so if we directly support Arrays, then I think we'll also have to put some thought into supporting other types as well.

How to "type" the event data is something that I think we will have to figure out eventually, but for now I think Strings are ideal.